File: src/main/java/com/kingmang/lazurite/parser/AST/Statements/UsingStatement.java
package com.kingmang.lazurite.parser.AST.Statements;

import com.kingmang.lazurite.exceptions.LzrException;

import com.kingmang.lazurite.parser.AST.Expressions.Expression;
import com.kingmang.lazurite.parser.AST.InterruptableNode;
import com.kingmang.lazurite.parser.parse.Parser;
import com.kingmang.lazurite.parser.parse.Token;
import com.kingmang.lazurite.parser.parse.classes.LexerImplementation;
import com.kingmang.lazurite.parser.parse.classes.ParserImplementation;
import com.kingmang.lazurite.patterns.visitor.FunctionAdder;
import com.kingmang.lazurite.patterns.visitor.ResultVisitor;
import com.kingmang.lazurite.patterns.visitor.Visitor;
import com.kingmang.lazurite.core.Types;
import com.kingmang.lazurite.runner.Runner;
import com.kingmang.lazurite.runner.RunnerInfo;
import com.kingmang.lazurite.runtime.values.LzrString;
import com.kingmang.lazurite.runtime.values.LzrValue;
import com.kingmang.lazurite.libraries.Library;
import com.kingmang.lazurite.utils.Loader;
import lombok.RequiredArgsConstructor;

import java.io.IOException;
import java.net.MalformedURLException;
import java.net.URL;
import java.net.URLClassLoader;
import java.util.List;


@RequiredArgsConstructor
public final class UsingStatement extends InterruptableNode implements Statement {

    private static final String PACKAGE = "com.kingmang.lazurite.libraries.%s.%s.%s.%s";

    
    public final Expression expression;

    @Override
    public void execute() {
        super.interruptionCheck();
        final LzrValue value = expression.eval();
        String[] parts = value.asString().split("\\.");
        //load Lzr file
        try {
            try {

                final Statement program = loadLzrLibrary(value.asString());
                program.accept(new FunctionAdder());
                program.execute();
            } catch (Exception ex) {
                throw new RuntimeException(ex);
            }

        }catch (Exception e){
            //load Lzr libs
            try{
                String libPackage = parts[0];

                try {
                    String libName = parts[1];
                    loadModule(libPackage, libName);

                }catch (Exception libEx){
                    String libChild = parts[1];
                    String libName = parts[2];
                    loadModule(libPackage, libChild, libName);
                }
            //load .jar libs
            }catch (Exception libEx){
                String[] jarParts = value.asString().split("::");
                String nameOfLib = jarParts[0] + ".jar";
                String pathToLib = jarParts[1];
                try {
                    URLClassLoader child = new URLClassLoader(
                            new URL[] { new URL("file:" + RunnerInfo.getPathToLzrLibs() + nameOfLib) },
                            this.getClass().getClassLoader()
                    );
                    Library module;
                    try {
                        module = (Library) Class.forName(pathToLib, true, child).newInstance();
                    } catch (ClassNotFoundException ex) {
                        module = (Library) Class.forName(pathToLib + ".invoker", true, child).newInstance();
                    }
                    module.init();
                } catch (MalformedURLException | InstantiationException | IllegalAccessException | ClassNotFoundException ex) {
                    throw new LzrException(ex.toString(), ex.getMessage());
                }
            }
        }
    }

    private void loadModule(String libPackage, String libChild, String name) {
        try {
            final Library module = (Library) Class.forName(String.format(PACKAGE, libPackage, libChild,name, name)).newInstance();
            module.init();
        } catch (Exception ex) {
            throw new LzrException("RuntimeException", "Unable to load module " + name + "\n" + ex);


        }
    }
    public Statement loadLzrLibrary(String path) throws IOException {
        final String input = Loader.readSource(path);
        //String input = "func danu(){print(111)}";
        final List<Token> tokens = LexerImplementation.tokenize(input);
        final Parser parser = new ParserImplementation(tokens);
        final Statement program = parser.parse();
        if (parser.getParseErrors().hasErrors()) {
            throw new LzrException("ParseException ", parser.getParseErrors().toString());
        }
        return program;
    }

    private void loadModule(String libPackage, String name) {
            final Library module;
            try {
                module = (Library) Class.forName(String.format("com.kingmang.lazurite.libraries.%s.%s.%s", libPackage,name, name)).newInstance();
                module.init();
            } catch (Exception e) {
                throw new LzrException("RuntimeException", "Unable to load module " + name + "\n" + e);
            }
    }
    @Override
    public void accept(Visitor visitor) {
        visitor.visit(this);
    }

    @Override
    public <R, T> R accept(ResultVisitor<R, T> visitor, T t) {
        return visitor.visit(this, t);
    }

    @Override
    public String toString() {
        return "using " + expression;
    }
}

File: src/main/java/com/kingmang/lazurite/parser/ast/statements/UsingStatement.kt
package com.kingmang.lazurite.parser.ast.statements

import com.kingmang.lazurite.exceptions.LzrException
import com.kingmang.lazurite.libraries.Library
import com.kingmang.lazurite.parser.IParser
import com.kingmang.lazurite.parser.ast.InterruptableNode
import com.kingmang.lazurite.parser.ast.expressions.Expression
import com.kingmang.lazurite.parser.impl.LexerImplementation
import com.kingmang.lazurite.parser.impl.ParserImplementation
import com.kingmang.lazurite.patterns.visitor.FunctionAdder
import com.kingmang.lazurite.patterns.visitor.ResultVisitor
import com.kingmang.lazurite.patterns.visitor.Visitor
import com.kingmang.lazurite.runner.RunnerInfo
import com.kingmang.lazurite.runtime.Libraries.add
import com.kingmang.lazurite.runtime.Libraries.isExists
import com.kingmang.lazurite.utils.Loader.readSource
import java.io.IOException
import java.net.MalformedURLException
import java.net.URL
import java.net.URLClassLoader

data class UsingStatement(val expression: Expression) : InterruptableNode(), Statement {
    override fun execute() {
        super.interruptionCheck()
        val value = expression.eval()
        val parts = value.asString().split("\\.".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        //load Lzr file
        try {
            try {
                val file = value.asString()
                // Check
                if (isExists(file))
                    return
                add(file)
                // Load & Run
                val program = loadLzrLibrary(file)
                program.accept(FunctionAdder())
                program.execute()
            } catch (ex: Exception) {
                throw RuntimeException(ex)
            }
        } catch (e: Exception) {
            //load Lzr libs
            try {
                val libPackage = parts[0]

                try {
                    val libName = parts[1]
                    loadModule(libPackage, libName)
                } catch (libEx: Exception) {
                    val libChild = parts[1]
                    val libName = parts[2]
                    loadModule(libPackage, libChild, libName)
                }
                //load .jar libs
            } catch (libEx: Exception) {
                val jarParts = value.asString().split("::".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                val nameOfLib = jarParts[0] + ".jar"
                val pathToLib = jarParts[1]
                try {
                    val child = URLClassLoader(arrayOf(URL("file:" + RunnerInfo.LZR_LIBS_PATH + nameOfLib)), this.javaClass.classLoader)
                    val module =
                        try {
                            Class.forName(pathToLib, true, child).newInstance() as Library
                        } catch (ex: ClassNotFoundException) {
                            Class.forName("$pathToLib.invoker", true, child).newInstance() as Library
                        }
                    module.init()
                } catch (ex: MalformedURLException) {
                    throw LzrException(ex.toString(), ex.message!!)
                } catch (ex: InstantiationException) {
                    throw LzrException(ex.toString(), ex.message!!)
                } catch (ex: IllegalAccessException) {
                    throw LzrException(ex.toString(), ex.message!!)
                } catch (ex: ClassNotFoundException) {
                    throw LzrException(ex.toString(), ex.message!!)
                }
            }
        }
    }

    private fun loadModule(libPackage: String, libChild: String, name: String) {
        try {
            val module = Class.forName(String.format(PACKAGE, libPackage, libChild, name, name)).newInstance() as Library
            module.init()
        } catch (ex: Exception) {
            throw LzrException("RuntimeException", "Unable to load module $name\n$ex")
        }
    }

    @Throws(IOException::class)
    fun loadLzrLibrary(path: String?): Statement {
        val input = readSource(path!!)
        val tokens = LexerImplementation.tokenize(input)
        val parser: IParser = ParserImplementation(tokens, path)
        val program = parser.parse()
        if (parser.parseErrors.hasErrors())
            throw LzrException("ParseException ", parser.parseErrors.toString())
        return program
    }

    private fun loadModule(libPackage: String, name: String) {
        try {
            val module = Class.forName(String.format(PACKAGE, libPackage, name, name)).newInstance() as Library
            module.init()
        } catch (e: Exception) {
            throw LzrException("RuntimeException", "Unable to load module $name\n$e")
        }
    }

    override fun accept(visitor: Visitor) =
        visitor.visit(this)

    override fun <R, T> accept(visitor: ResultVisitor<R, T>, input: T): R? =
        visitor.visit(this, input)

    override fun toString(): String =
        "using ${this.expression}"

    companion object {
        private const val PACKAGE = "com.kingmang.lazurite.libraries.%s.%s.%s.%s"
    }
}
